### PR TITLE
fix(pipeline): pass termination-aware discount to HordeActorCriticAgent in horde_ac control mode

### LINKED ISSUE #2344: AlbertaPipeline control_mode="horde_ac" drops `terminated`: the value head bootstraps and the actor trace survives every episode boundary
`AlbertaPipeline.update` documents a `terminated` argument ("Scalar termination
flag (`0.0` or `1.0`)") and validates it as a float32 scalar restricted to
`{0, 1}`. On the default `control_mode="sarsa"` path it is forwarded into
`step4_update` → `SARSAAgent.update`, where it skips bootstrapping
(`skip_bootstrap = (safe_terminated != 0) | (gamma_arr == 0.0)`,
`alberta_framework/core/sarsa.py:945`) and zeroes the eligibility traces
(`sarsa.py:1005-1006`).

On the registered `control_mode="horde_ac"` path it is dropped on the floor.
`alberta_framework/pipeline.py:1576-1581` calls

```python
ac_result = ac.update(
    ac_state,
    reward,
    features,
    auxiliary_cumulants=auxiliary_cumulants,
)
```

with no `discount`, so `HordeActorCriticAgent.update` falls back to
`value_discount = self._critic.horde_spec.gammas[cfg.value_head_index]`
(`alberta_framework/core/horde_actor_critic.py:1001-1003`) on *every* step,
terminal or not.

Two things are consequently wrong at a terminal transition:

1. **The value head bootstraps through termination.** The critic's TD target
   for the value demon is `r + gamma * V(s')` instead of `r`.
2. **The actor eligibility trace survives the boundary.** In
   `HordeActorCriticAgent.update`, `actor_decay = value_discount * cfg.actor_lamda`
   and `carry_traces = value_discount != 0.0` (`horde_actor_critic.py:1044,1061`),
   so with `discount = 0` the actor trace is zeroed — but the pipeline never
   supplies it, and the trace is carried into the next episode instead.

The mechanism to fix it already exists and was built for exactly this caller.
`HordeActorCriticAgent.update` takes `discount: Array | None` — "Optional
scalar per-transition value-head discount" — and routes to
`HordeLearner.update_with_discounts`, whose docstring says it "lets control
adapters zero the value head at episode boundaries while keeping the Horde's
fixed GVF metadata and per-head trace decay intact"
(`alberta_framework/core/horde.py:554-557`). The pipeline is such a control
adapter and is the only in-tree one that does not use it.

## Reproduction

`alberta_framework/pipeline.py` at `c9aba7b5`, identity Step 2 features,
`horde.gammas=(0.9, 0.5)`, `value_head_index=0`, `actor_lamda=0.8`:

```python
state = pipeline.init(jr.key(0), jnp.asarray([0.2, -0.1, 0.4], dtype=jnp.float32))
obs = jnp.asarray([0.1, 0.3, -0.2], dtype=jnp.float32)
reward = jnp.asarray(0.5, dtype=jnp.float32)
cumulants = jnp.asarray([0.5, -0.2], dtype=jnp.float32)
for term in (0.0, 1.0):
    res = pipeline.update(state, obs, reward, jnp.asarray(term, dtype=jnp.float32), cumulants)
    print(term, res.horde_td_targets, res.state.control_state.actor_trace_weights)
```

```
0.0 [ 0.34537804 -0.18965168] [[ 0.1 -0.05  0.2] [-0.1  0.05 -0.2]]
1.0 [ 0.34537804 -0.18965168] [[ 0.1 -0.05  0.2] [-0.1  0.05 -0.2]]
```

The two lines are byte-identical: `terminated` has **no effect whatsoever** on
this control path. The value-head target should be `0.5` (the reward alone) in
the terminal case; it is `0.5 + 0.9 * V(s') = 0.5 + 0.9 * (-0.1718) = 0.3454`.
The actor trace should be empty; it is not.

## Why no test caught it

Every existing pipeline test passes `terminated = 0.0`
(`tests/test_pipeline.py:237`, `:280`, `:324`, `:412`, `:484`, `:540`, `:580`;
the array-run helper builds `terminated = jnp.zeros(n_steps)` at `:632`). `terminated = 1.0` is exercised nowhere in
the pipeline suite, on either control mode, so the entire termination semantics
of the documented flag are untested and the `horde_ac` branch's untested path
is silently wrong.

## Scope

`AlbertaPipeline.run_arrays` delegates to `AlbertaPipeline.update`, so array
runs and `run_pipeline_smoke` inherit the same defect. Any measurement taken
through the pipeline with `control_mode="horde_ac"` on an episodic stream
learns terminal transitions as if the episode continued. Auxiliary GVF demons
are *not* affected and must keep their own configured gammas — only the
value head's discount is a per-transition control quantity.

## Proposed fix

Pass the terminal-aware discount in the `horde_ac` branch:

```python
value_gamma = self._horde.horde_spec.gammas[value_index]
control_discount = jnp.where(terminated == 0.0, value_gamma, jnp.zeros_like(value_gamma))
ac_result = ac.update(..., discount=control_discount)
```

At `terminated == 0` this is a bit-exact reduction: `update_with_discounts`
computes `bootstrap = jnp.where(discounts == 0.0, 0.0, discounts * next_preds)`
from the same `gammas` vector the `discount is None` branch uses, and the extra
finite/range head-validity conditions are satisfied by the configured gammas.

Lane/metric: this is a measurement-trustworthiness fix, not a benchmark climb —
no `outputs/` artifact changes and no registered evidence source is touched.

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2344

## Summary
In `alberta_framework/pipeline.py`:
When `control_mode == "horde_ac"`, `AlbertaPipeline.update` called `ac.update(...)` without forwarding the `terminated` argument (or passing `discount`). As a result, `HordeActorCriticAgent.update` defaulted to using the fixed value head demon gamma `self._critic.horde_spec.gammas[value_head_index]` on every step, causing the value head to bootstrap through episode terminations ($r + \gamma V(s')$ instead of $r$) and carrying the actor's eligibility trace across episode boundaries.

## Proposed Changes
1. Computed `control_discount = jnp.where(terminated == 0.0, value_gamma, jnp.zeros_like(value_gamma))`.
2. Passed `discount=control_discount` to `ac.update(...)` in the `horde_ac` control branch of `AlbertaPipeline.update`.
3. Added unit tests in `tests/test_pipeline.py` verifying that when `terminated == 1.0`, the value head target equals the reward alone and actor eligibility traces are reset.

## Verification
- Ran pytest: `/opt/homebrew/bin/uv run pytest -o pythonpath=. tests/test_pipeline.py` (184/184 passed).
- Ran ruff: `/opt/homebrew/bin/uv run ruff check alberta_framework/pipeline.py tests/test_pipeline.py` (clean).
- Ran mypy: `/opt/homebrew/bin/uv run mypy --follow-imports=silent alberta_framework/pipeline.py` (clean).

### CODE REVIEW

**Review by lumix5 — APPROVED** [review]
Independent verification at the current head (no prior review existed at any head).

**Disposition: merge.**

- **The wiring gap is real:** `HordeActorCritic.update` has accepted an explicit `discount` parameter all along (horde_actor_critic.py:965), but the `horde_ac` branch of `pipeline.update` never passed one — so the control learner bootstrapped off the post-terminal value estimate and never zeroed actor traces at episode boundaries, while the SARSA control path right below it does handle `terminated` explicitly (via `step4_update`).
- **The fix passes exactly the right quantity:** `where(terminated == 0, horde_spec.gammas[value_head_index], 0)` — the value head's own gamma during continuation, zero at termination, which is the standard terminal TD(λ) treatment the QHordeActorCritic sibling already implements internally (`effective_gamma = where(terminated, 0, gamma)`).
- **Ran the full patched `tests/test_pipeline.py` (184 tests) in a throwaway container** (network denied, branch `pipeline.py` overlaid on unchanged main): all pass, including the new terminal-semantics test asserting the value head's TD target collapses to the bare reward (0.5, not the bootstrapped 0.3454) and actor trace weights zero out at termination while the continuing step keeps both distinct. ci-ok green at head.

Non-blocking note: `value_gamma` is `horde_spec.gammas[value_index]` while the actor lambda remains config-bound — consistent with the demon-discount semantics elsewhere in the pipeline, just worth a glance when someone next touches the actor trace discounting question.

**Comment by lalalune** [comment]
Reviewed: Superseded by #2348 after comparing diffs: same terminal discount correction; retained version tests nonterminal behavior, auxiliary heads and traces, without unrelated formatting churn.
